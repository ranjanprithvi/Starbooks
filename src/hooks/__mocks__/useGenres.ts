const useGenres = jest.fn().mockReturnValue({
    genres: [
        { _id: "1", name: "genre1" },
        { _id: "2", name: "genre2" },
    ],
    setGenres: jest.fn(),
    error: "",
    setError: jest.fn(),
    isLoading: false,
});

export default useGenres;
