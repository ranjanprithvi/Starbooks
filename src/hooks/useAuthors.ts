import { Author } from "../models/author";
import useData from "./generic/useData";

const useAuthors = () => {
    const {
        data: authors,
        setData: setAuthors,
        error,
        setError,
        isLoading,
    } = useData<Author>("/authors");
    return { authors, setAuthors, error, setError, isLoading };
};

export default useAuthors;
