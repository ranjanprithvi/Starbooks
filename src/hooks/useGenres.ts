import { Genre } from "../models/genre";
import useData from "./generic/useData";

const useGenres = () => {
    const {
        data: genres,
        setData: setGenres,
        error,
        setError,
        isLoading,
    } = useData<Genre>("/genres");
    return { genres, setGenres, error, setError, isLoading };
};

export default useGenres;
