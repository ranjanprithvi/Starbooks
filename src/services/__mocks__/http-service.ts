export const mockPost = jest.fn();
export const mockPatch = jest.fn();
export const mockDelete = jest.fn();
export const mockGet = jest.fn();
export const mockGetAll = jest.fn();

const MockedHttpService = jest.fn().mockImplementation(() => {
    return {
        post: mockPost,
        patch: mockPatch,
        delete: mockDelete,
        get: mockGet,
        getAll: mockGetAll,
    };
});

export default MockedHttpService;
