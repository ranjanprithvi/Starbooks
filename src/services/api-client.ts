import axios, { CanceledError } from "axios";

export default axios.create({
    params: {
        // key:"value"
    },
    headers: { "x-auth-token": localStorage.getItem("token") },
    baseURL: "https://starbooks-backend.onrender.com/api",
});

export { CanceledError };
