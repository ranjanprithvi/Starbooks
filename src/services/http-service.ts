import { AxiosRequestConfig } from "axios";
import apiClient from "./api-client";

export interface Entity {
    _id: string;
}

class HttpService<T extends Entity> {
    endpoint: string;

    constructor(endpoint: string) {
        this.endpoint = endpoint;
    }

    getAll(query: Object | null) {
        const controller = new AbortController();
        return {
            request: apiClient.get<T[]>(this.endpoint, {
                signal: controller.signal,
                params: query,
            }),
            cancel: () => controller.abort(),
        };
    }

    delete(id: number) {
        return apiClient.delete(`${this.endpoint}/${id}`);
    }

    add(entity: T) {
        return apiClient.post<T>(this.endpoint, entity);
    }

    update(entity: T) {
        return apiClient.put(`${this.endpoint}/${entity._id}`, entity);
    }
}

export default HttpService;
