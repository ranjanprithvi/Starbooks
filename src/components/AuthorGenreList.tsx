import useGenres from "../hooks/useGenres";
import useAuthors from "../hooks/useAuthors";
import {
    Button,
    GridItem,
    HStack,
    Heading,
    Input,
    List,
    ListItem,
    MenuItem,
    VStack,
    useToast,
} from "@chakra-ui/react";
import { useRef } from "react";
import HttpService from "../services/http-service";

const AuthorGenreList = () => {
    const { genres } = useGenres();
    const { authors } = useAuthors();
    const toast = useToast();

    const lists = [
        {
            heading: "Authors",
            name: "author",
            items: authors,
            inputRef: useRef(null),
        },
        {
            heading: "Genres",
            name: "genre",
            items: genres,
            inputRef: useRef(null),
        },
    ];

    const onAdd = (listName: string, value: string = "") => {
        const service = new HttpService(`/${listName}s`);

        service
            .post({ name: value })
            .then(() => {
                window.location.reload();
            })
            .catch((err) => {
                toast({
                    title: "Error",
                    description: err.response.data,
                    status: "error",
                    duration: 5000,
                    isClosable: true,
                });
            });
    };

    return (
        <GridItem colSpan={2} width="auto" margin={5}>
            <HStack justifyContent="space-evenly">
                {lists.map((list) => (
                    <VStack
                        key={list.name}
                        border="2px"
                        borderRadius={10}
                        padding={10}
                    >
                        <Heading> {list.heading}</Heading>
                        <List
                            width="100%"
                            height="50vh"
                            overflow="scroll"
                            border="1px"
                            borderColor="gray.200"
                        >
                            {list.items.map((item) => (
                                <ListItem
                                    paddingY="2"
                                    paddingRight="10"
                                    paddingLeft="5"
                                    key={item._id}
                                    border="1px"
                                    borderColor="gray.50"
                                >
                                    {item.name}
                                </ListItem>
                            ))}
                        </List>
                        <HStack marginY="2">
                            <Input
                                ref={list.inputRef}
                                placeholder={`Enter new ${list.name}..`}
                            />
                            <Button
                                colorScheme="teal"
                                paddingX="8"
                                onClick={() => {
                                    const textBox = list.inputRef.current
                                        ? (list.inputRef
                                              .current as HTMLInputElement)
                                        : null;
                                    onAdd(list.name, textBox?.value);
                                }}
                            >
                                Add {list.name}
                            </Button>
                        </HStack>
                    </VStack>
                ))}
            </HStack>
        </GridItem>
    );
};

export default AuthorGenreList;
