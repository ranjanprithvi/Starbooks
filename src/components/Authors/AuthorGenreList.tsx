import useGenres from "../../hooks/useGenres";
import useAuthors from "../../hooks/useAuthors";
import {
    Box,
    Button,
    Flex,
    GridItem,
    HStack,
    Heading,
    Input,
    List,
    ListItem,
    MenuItem,
    VStack,
    useToast,
} from "@chakra-ui/react";
import { useRef } from "react";
import { httpService } from "../../services/http-service";

const AuthorGenreList = () => {
    const { genres } = useGenres();
    const { authors } = useAuthors();
    const toast = useToast();

    const lists = [
        {
            heading: "Authors",
            name: "author",
            items: authors,
            inputRef: useRef(null),
        },
        {
            heading: "Genres",
            name: "genre",
            items: genres,
            inputRef: useRef(null),
        },
    ];

    const handleAdd = (listName: string, value: string = "") => {
        const service = httpService(`/${listName}s`);

        service
            .post({ name: value })
            .then(() => {
                window.location.reload();
            })
            .catch((err) => {
                toast({
                    title: "Error",
                    description:
                        err.response?.data || "Sorry. Something went wrong",
                    status: "error",
                    duration: 5000,
                    isClosable: true,
                });
            });
    };

    const handleDelete = (listName: string, id: string = "") => {
        const service = httpService(`/${listName}s`);

        service
            .delete(id)
            .then(() => {
                window.location.reload();
            })
            .catch((err) => {
                toast({
                    title: "Error",
                    description:
                        err.response?.data || "Sorry. Something went wrong",
                    status: "error",
                    duration: 5000,
                    isClosable: true,
                });
            });
    };

    return (
        <GridItem colSpan={2} width="auto" marginX="5">
            <Flex
                direction={{ base: "column", md: "row" }}
                justifyContent="space-evenly"
                gap={5}
            >
                {lists.map((list) => (
                    <VStack
                        key={list.name}
                        border="2px"
                        borderRadius={10}
                        padding={10}
                    >
                        <Heading> {list.heading}</Heading>
                        <List
                            width="100%"
                            height="50vh"
                            overflow="scroll"
                            border="1px"
                            borderColor="gray.200"
                        >
                            {list.items.map((item) => (
                                <ListItem
                                    paddingY="2"
                                    paddingRight="10"
                                    paddingLeft="5"
                                    key={item._id}
                                    border="1px"
                                    borderColor="gray.50"
                                    display="flex"
                                    justifyContent="space-between"
                                >
                                    <Box>{item.name}</Box>
                                    <Button
                                        data-testid={`delete-${list.name}-${item._id}`}
                                        size="sm"
                                        onClick={() =>
                                            handleDelete(list.name, item._id)
                                        }
                                    >
                                        -
                                    </Button>
                                </ListItem>
                            ))}
                        </List>
                        <HStack marginY="2">
                            <Input
                                ref={list.inputRef}
                                placeholder={`Enter new ${list.name}..`}
                            />
                            <Button
                                data-testid={`add-${list.name}-button`}
                                colorScheme="teal"
                                paddingX="8"
                                onClick={() => {
                                    const textBox = list.inputRef.current
                                        ? (list.inputRef
                                              .current as HTMLInputElement)
                                        : null;
                                    handleAdd(list.name, textBox?.value);
                                }}
                            >
                                Add {list.name}
                            </Button>
                        </HStack>
                    </VStack>
                ))}
            </Flex>
        </GridItem>
    );
};

export default AuthorGenreList;
