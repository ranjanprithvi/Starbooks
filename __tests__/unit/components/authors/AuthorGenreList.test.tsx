import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import AuthorGenreList from "../../../../src/components/Authors/AuthorGenreList";
import useAuthors from "../../../../src/hooks/useAuthors";
import useGenres from "../../../../src/hooks/useGenres";
import "@testing-library/jest-dom";
import { httpService } from "../../../../src/services/http-service";

//mock window.location.replace
const mReload = jest.fn();
Object.defineProperty(window, "location", {
    value: {
        reload: mReload,
    },
});

// mock HttpService
jest.mock("../../../../src/services/http-service");
const mHttpService = httpService as jest.MockedFunction<typeof httpService>;
const mPost = jest.fn();
mPost.mockResolvedValue({ data: { _id: "3", name: "genre3" } });
const mDelete = jest.fn();
mDelete.mockResolvedValue({ data: { _id: "3", name: "genre3" } });

mHttpService.mockReturnValue({
    post: mPost,
    delete: mDelete,
    get: jest.fn(),
    patch: jest.fn(),
} as any);
// const mHttpService = HttpService as jest.Mocked<typeof HttpService>;

// beforeEach(() => {
//     HttpService.mockClear();
// });

// const mHttpObj = jest.fn();
// mHttpService.mockImplementation(() => {});

//mock useAuthors
jest.mock("../../../../src/hooks/useAuthors");
const mUseAuthors = useAuthors as jest.MockedFunction<typeof useAuthors>;

//mock useAuthors
jest.mock("../../../../src/hooks/useGenres");
const mUseGenres = useGenres as jest.MockedFunction<typeof useGenres>;

beforeAll(() => {
    mUseAuthors.mockReturnValue({
        authors: [
            {
                _id: "1",
                name: "author1",
            },
            {
                _id: "2",
                name: "author2",
            },
        ],
        error: "",
        setAuthors: jest.fn(),
        setError: jest.fn(),
        isLoading: false,
    });

    mUseGenres.mockReturnValue({
        genres: [
            {
                _id: "1",
                name: "genre1",
            },
            {
                _id: "2",
                name: "genre2",
            },
        ],
        error: "",
        setGenres: jest.fn(),
        setError: jest.fn(),
        isLoading: false,
    });
});

beforeEach(() => {
    jest.clearAllMocks();
});

describe("AuthorGenreList", () => {
    it("renders correctly", () => {
        const tree = renderer.create(<AuthorGenreList />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("calls post method of HttpService when add button is clicked", async () => {
        render(<AuthorGenreList />);
        act(() => {
            fireEvent.change(
                screen.getByPlaceholderText(/enter new genre../i),
                {
                    target: { value: "genre3" },
                }
            );
        });
        fireEvent.click(screen.getByTestId("add-genre-button"));
        expect(mHttpService).toHaveBeenCalledTimes(1);
        expect(mPost).toHaveBeenCalledWith({ name: "genre3" });
        await waitFor(() => expect(mReload).toHaveBeenCalledTimes(1));
    });

    it("calls post method of HttpService when delete button is clicked", async () => {
        render(<AuthorGenreList />);

        fireEvent.click(screen.getByTestId("delete-genre-1"));
        expect(mHttpService).toHaveBeenCalledTimes(1);
        expect(mDelete).toHaveBeenCalledWith("1");
        await waitFor(() => expect(mReload).toHaveBeenCalledTimes(1));
    });
});
