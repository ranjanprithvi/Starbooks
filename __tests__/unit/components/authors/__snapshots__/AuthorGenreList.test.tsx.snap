// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorGenreList renders correctly 1`] = `
<div
  className="css-v3bkq2"
>
  <div
    className="css-qaksvd"
  >
    <div
      className="chakra-stack css-12gdbjn"
    >
      <h2
        className="chakra-heading css-0"
      >
         
        Authors
      </h2>
      <ul
        className="css-152zwno"
        role="list"
      >
        <li
          className="css-cs8ahz"
        >
          <div
            className="css-0"
          >
            author1
          </div>
          <button
            className="chakra-button css-4xx2wk"
            data-testid="delete-author-1"
            onClick={[Function]}
            type="button"
          >
            -
          </button>
        </li>
        <li
          className="css-cs8ahz"
        >
          <div
            className="css-0"
          >
            author2
          </div>
          <button
            className="chakra-button css-4xx2wk"
            data-testid="delete-author-2"
            onClick={[Function]}
            type="button"
          >
            -
          </button>
        </li>
      </ul>
      <div
        className="chakra-stack css-v3ng2i"
      >
        <input
          className="chakra-input css-0"
          onBlur={[Function]}
          onFocus={[Function]}
          placeholder="Enter new author.."
        />
        <button
          className="chakra-button css-kl83lb"
          data-testid="add-author-button"
          onClick={[Function]}
          type="button"
        >
          Add 
          author
        </button>
      </div>
    </div>
    <div
      className="chakra-stack css-12gdbjn"
    >
      <h2
        className="chakra-heading css-0"
      >
         
        Genres
      </h2>
      <ul
        className="css-152zwno"
        role="list"
      >
        <li
          className="css-cs8ahz"
        >
          <div
            className="css-0"
          >
            genre1
          </div>
          <button
            className="chakra-button css-4xx2wk"
            data-testid="delete-genre-1"
            onClick={[Function]}
            type="button"
          >
            -
          </button>
        </li>
        <li
          className="css-cs8ahz"
        >
          <div
            className="css-0"
          >
            genre2
          </div>
          <button
            className="chakra-button css-4xx2wk"
            data-testid="delete-genre-2"
            onClick={[Function]}
            type="button"
          >
            -
          </button>
        </li>
      </ul>
      <div
        className="chakra-stack css-v3ng2i"
      >
        <input
          className="chakra-input css-0"
          onBlur={[Function]}
          onFocus={[Function]}
          placeholder="Enter new genre.."
        />
        <button
          className="chakra-button css-kl83lb"
          data-testid="add-genre-button"
          onClick={[Function]}
          type="button"
        >
          Add 
          genre
        </button>
      </div>
    </div>
  </div>
</div>
`;
